core/test/mpi/base/communicator.cpp

```diff
@@ -0,0 +1,172 @@
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <mpi.h>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/mpi.hpp>
+
+
+namespace {
+
+
+class Communicator : public ::testing::Test {
+protected:
+    Communicator() : comm(MPI_COMM_WORLD) {}
+
+    void SetUp()
+    {
+        rank = comm.rank();
+        ASSERT_EQ(comm.size(), 8);
+    }
+
+    gko::mpi::communicator comm;
+    int rank;
+};
+
+
+TEST_F(Communicator, CommKnowsItsSize)
+{
+    int size;
+    MPI_Comm_size(MPI_COMM_WORLD, &size);
+
+    EXPECT_EQ(comm.size(), size);
+}
+
+
+TEST_F(Communicator, CommKnowsItsRank)
+{
+    int rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    EXPECT_EQ(comm.rank(), rank);
+}
+
+
+TEST_F(Communicator, CommKnowsItsLocalRank)
+{
+    int rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    // Expect local rank to be same as rank when on one node
+    EXPECT_EQ(comm.node_local_rank(), rank);
+}
+
+
+TEST_F(Communicator, CommunicatorCanBeCopyConstructed)
+{
+    gko::mpi::communicator copy(comm);
+
+    EXPECT_EQ(copy == comm, true);
+}
+
+
+TEST_F(Communicator, CommunicatorCanBeSynchronized)
+{
+    ASSERT_NO_THROW(comm.synchronize());
+}
+
+
+TEST_F(Communicator, CommunicatorCanBeCopyAssigned)
+{
+    gko::mpi::communicator copy = comm;
+
+    EXPECT_EQ(copy == comm, true);
+}
+
+
+TEST_F(Communicator, NonOwnedCommunicatorFailsToMove)
+{
+    ASSERT_THROW({ auto moved = std::move(comm); }, gko::NotSupported);
+}
+
+
+TEST_F(Communicator, NonOwnedCommunicatorFailsToMoveConstruct)
+{
+    ASSERT_THROW({ auto moved = gko::mpi::communicator(std::move(comm)); },
+                 gko::NotSupported);
+}
+
+
+TEST_F(Communicator, CanSetCustomCommunicator)
+{
+    auto world_rank = comm.rank();
+    auto world_size = comm.size();
+    auto color = world_rank / 4;
+
+    auto row_comm = gko::mpi::communicator(comm.get(), color, world_rank);
+    for (auto i = 0; i < world_size; ++i) {
+        EXPECT_LT(row_comm.rank(), 4);
+    }
+}
+
+
+TEST_F(Communicator, CanMoveAssignCustomCommunicator)
+{
+    auto world_rank = comm.rank();
+    auto world_size = comm.size();
+    auto color = world_rank / 4;
+
+    auto row_comm = gko::mpi::communicator(comm.get(), color, world_rank);
+    auto mv_row_comm = std::move(row_comm);
```

Comment:
This calls the move constructor. You probably want
```suggestion
    gko::mpi::communicator mv_row_comm(MPI_COMM_WORLD);
    mv_row_comm = std::move(row_comm);
```